packages/editor/src/lib/app/App.ts

```diff
@@ -211,14 +211,28 @@ export class App extends EventEmitter<TLEventMap> {
 			})
 		}
 
+		let deletedShapes: TLShape[] | undefined = undefined
+
 		this.store.onBeforeDelete = (record) => {
-			if (record.typeName === 'shape') {
-				this._shapeWillBeDeleted(record)
-			} else if (record.typeName === 'page') {
-				this._pageWillBeDeleted(record)
+			switch (record.typeName) {
+				case 'shape': {
+					if (deletedShapes === undefined) {
+						deletedShapes = []
+					}
+					deletedShapes.push(record)
+					break
+				}
+				case 'page': {
+					this._pageWillBeDeleted(record)
+					break
+				}
 			}
 		}
 
+		if (deletedShapes !== undefined) {
+			this._shapesWillBeDeleted(deletedShapes)
+		}
```

Comment:
might be wrong, but I don't _think_ this does what you want it to? We're in the App constructor here so afaict the order will be

1. set `deletedShapes` to `undefined`
2. register the `onBeforeDelete` handler
3. check if `deletedShapes` is still `undefined`. As we're still in the constructor, there hasn't been any opportunity for `onBeforeDelete` to get called, so it always will be

I think with this change now `_shapesWillBeDeleted` just won't get called?

Reply:
This was a little sketch that I shared with David this morning; he pointed out the same thing! The point was to illustrate that if we called `_shape(s)WillBeDeleted` at the end of a transaction, that was much faster than calling it once per shape being deleted. But either way, I think he has a better solution. Closing!